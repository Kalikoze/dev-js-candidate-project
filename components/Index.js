import React, { Component } from 'react';
import fetch from 'isomorphic-unfetch';
import ChatContainer from '../containers/ChatContainer';
import { fetchRandomJoke, fetchNerdyJoke, fetchExplicitJoke } from '../store';
import Message from '../components/Message';
import Layout from '../components/Layout';
import PlayControls from '../components/PlayControls';

export default class Index extends Component {
  constructor() {
    super();
    this.state = {
      messages: [],
    };
    this.addMessage = this.addMessage.bind(this);
  }

  static async getInitialProps({ query }) {
    return query;
  }

  componentDidMount() {
    const { messages } = this.state;
    const { message } = this.props;
    if(this.props.message) {
      this.setState({messages: [...messages, {message, isUser: false}]})
    }
  }

  componentDidUpdate() {
    this.convo.scrollTop = this.convo.scrollHeight;
  }

  async addMessage(message, intent) {
    const { messages } = this.state;
    const { fetchRandomJoke, fetchNerdyJoke, fetchExplicitJoke } = this.props.fetchJoke;
    const jokeIntents = ['randomjoke', 'nerdyjoke', 'explicitjoke'];
    const getJoke = [fetchRandomJoke, fetchNerdyJoke, fetchExplicitJoke];

    if(jokeIntents.includes(intent)) {
      const index = jokeIntents.indexOf(intent);
      await getJoke[index]();

      return this.setState({messages: [...messages, {message: `${message} ${this.props.currentJoke}`, isUser: !intent ? true : false }]});
    }

    this.setState({messages: [...messages, {message, isUser: !intent ? true : false }]});
  }

  render() {
    const { messages } = this.state;
    const conversation = messages.map((message, i) => <Message key={i} text={message.message} isUser={message.isUser}/>);
    return (
      <Layout>
        <section className='header'>
          <h1>SpruceBot</h1>
          <PlayControls />
        </section>
        <section className='conversation' ref={node => this.convo = node}>
          {conversation}
        </section>
        <ChatContainer addMessage={this.addMessage}/>
        <style jsx>{`
          .header {
            align-items: center;
            color: #FFF;
            display: flex;
            font-family: 'Lato', sans-serif;
            font-size: 72px;
            font-weight: 700;
            height: 20%;
            justify-content: center;
            width: 100%;
          }

          .conversation {
            display: flex;
            flex-direction: column;
            height: 65%;
            overflow-y: scroll;
            width: 70%;
          }
        `}</style>
        <style jsx global>{`
          * {
            box-sizing: border-box;
            margin: 0;
            padding: 0;
          }
        `}</style>
      </Layout>
    );
  }
}
