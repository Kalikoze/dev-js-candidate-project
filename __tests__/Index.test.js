import React from 'react';
import Index from '../components/Index';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, configure } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
const fakeStore = configureMockStore()({watsonResponse: { response: ''}});

configure({ adapter: new Adapter() });

const setUp = () => {
  const wrapper = mount(
    <Provider store={fakeStore}>
      <Index />
    </Provider>
  );

  return { wrapper };
};

describe('Index page', () => {
  it('should render the Index page correctly with the right classes and styles', () => {
    const createNodeMock = (element) => {
      if (element.type === 'audio') {
        return {
          focus() {},
        };
      }
      return null;
    };

    const { wrapper } = setUp();

    const tree = renderer.create(wrapper, { createNodeMock }).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be defined', () => {
    const { wrapper } = setUp();

    expect(wrapper).toBeDefined();
  });

  it('should have an empty array set to messages in state', () => {
    const wrapper = shallow(<Index />);

    expect(wrapper.state().messages).toEqual([]);
  });
});
