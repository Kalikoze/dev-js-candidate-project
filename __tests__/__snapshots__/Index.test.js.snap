// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page should render the Index page correctly with the right classes and styles 1`] = `
<section
  className="jsx-2146746531 layout"
>
  <video
    autoPlay={true}
    className="jsx-2146746531"
    loop={true}
    src="../static/ambient-background.mp4"
  />
  <section
    className="jsx-2146746531 container"
  >
    <section
      className="jsx-3139470545 header"
    >
      <h1
        className="jsx-3139470545"
      >
        SpruceBot
      </h1>
      <section
        className="jsx-3144229325 play-controls"
      >
        <audio
          autoPlay={true}
          className="jsx-3144229325"
          loop={true}
          src="../static/music/track1.mp3"
        />
        <img
          className="jsx-3144229325"
          onClick={[Function]}
          src="../static/play-controls/rewind.svg"
        />
        <img
          className="jsx-3144229325"
          onClick={[Function]}
          src="../static/play-controls/pause-button.svg"
        />
        <img
          className="jsx-3144229325"
          onClick={[Function]}
          src="../static/play-controls/fast-forward.svg"
        />
      </section>
    </section>
    <section
      className="jsx-3139470545 conversation"
    />
    <section
      className="jsx-2439614921 input-section"
    >
      <input
        className="jsx-2439614921"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Chat with SpruceBot here..."
        value=""
      />
      <button
        className="jsx-2439614921"
        onClick={[Function]}
      />
    </section>
  </section>
</section>
`;
